### ISSUE TITLE
try_get?

### ISSUE THREAD

**Issue by quasicomputational** [bikeshedder/deadpool#14]
A `try_get` method (non-async, returning `Result<Option<T>, E>` would be nice to have if you would rather shed load than queue unboundedly and risk bufferbloat.

This is related to #9, but avoids the need to ever switch contexts or even touch the async machinery, and so does benefit from being in `deadpool` rather than user code.

**Comment by bikeshedder** [comment]
I don't know if this is actually feasible:
1. The pool starts empty thus `Pool::try_get` would always return `None` unless you ever create some objects beforehand.
2. The recycling happens inside `Pool::get` and calling it non-async would hand out a possibly non-usable object.
3. The objects are currently stored inside a `tokio::sync::mpsc::channel` which uses a `Mutex` around the `Receiver`. There is currently no way to `try_lock` the mutex or `try_recv` from receiver.

What could be possible though is an `try_get` that does not wait for an connection to become available but fail immediately if there are no connections are available and the pool size is already at its maximum.

**Comment by quasicomputational** [comment]
Ah, yeah, you're right: non-async `try_get` isn't possible at all. Even if the pool's internals allowed it, it'd have no way to use `create` or `recycle` on the manager!

That makes this idea a bit less of a win. I think it'd still be beneficial because it'd work in whatever runtime (as opposed to using a timeout from some specific runtime) and, again, doing less work.

**Comment by bikeshedder** [comment]
I'm happy to add a `try_get` (or something similar) to the `deadpool` API. What exactly are your needs? Those are the three things that the code is currently waiting for:

1. Create a new object

or

1. Wait for object to become available
2. Recycle object

I think the most obvious combination of a `try_get` would be to not wait for the object to become available but block on both creating the object and recycling it. But then it would definitely block if the database is reacting really slow or something else is going haywire. So I wonder if `get_timeout` with a configurable timeout for both object retrieval and creation/recycling would be a better idea?

**Comment by quasicomputational** [comment]
I agree with your analysis.

Really, what I want is an option to deal with excessive load - i.e., requests coming in faster than they can be served because of database contention - that isn't queueing indefinitely and hence potentially large latency. A timeout adds backpressure and cures a lot of the evils of queues, but a request still spends a little while stuck in a queue only to be told it won't be served, which isn't quite ideal.

`try_get` failing early when the pool's at capacity is the desired behaviour, because that lets me shed load quickly. Clients get told to back off early and spend less time waiting, and the server's consuming less resources because it doesn't need to maintain a queue at all.

That doesn't determine what it should do if there's not an object handy. But deadpool's (excellent) principle of treating startup and runtime the same reveals that `try_get` should block on creating an object, because otherwise the first iteration of the server's main loop breaks. Similarly, waiting on `recycle` is no problem and actually desirable - I *do* want the object, just not to be waiting on other requests to finish.

As for the limping database scenario: I think that's up to the `Manager` impl to detect as it creates/recycles a connection, right? I can already guard against this about as well as I can with a timeout in `create` or `recycle`; I don't see the win for deadpool to get involved here.

I can also bound the over-all latency with `timeout(pool.get())` just fine today, but there's not an easy way to separate a limping database (which is unavoidable latency) from stationary queues due to too-high load (which could be avoided by shedding load). Perhaps a timeout specifically for waiting on objects to be returned to the pool could be handy, so that users can pick their own place on the load-shedding-versus-queueing-latency trade-off. Maybe the ideal API here looks something like `try_get(timeout: Option<Duration>)`, with `None` meaning no timeout and queue indefinitely, `Some(0) meaning don't queue at all and shed load, and other values in between?

**Comment by bikeshedder** [comment]
I just submitted a PR to Tokio adding `Mutex::try_get` and `Receiver::try_recv` which should make it possilbe to implement this: https://github.com/tokio-rs/tokio/pull/1939

**Comment by bikeshedder** [comment]
@quasicomputational I just added support for timeouts and a non-blocking `Pool::try_get` function to deadpool. `0.4.0` is now released on crates.io. Could you please give this a try and tell me if there is still anything missing?
